### PR TITLE
Validators for relationships

### LINKED ISSUE #3903: Repository Handler is not checking if ends of relationships are null
"getOtherEnd" method from RepositoryHandler is not checking if the ends of relationships are not null.
This may cause a NPE. Before checking if the GUIDs are equal, a check for requiredEnd and startingEnd proxy should be added.

### PR DESCRIPTION
**Pull request by cmgrote**
Primarily to resolve #3903

### CODE REVIEW

**Comment by cmgrote** [comment]
@lpalashevski @danielaotelea can someone test that this at least now logs an issue and raises a RepositoryErrorException in such scenarios as documented in #3903 ?

**Comment by lpalashevski** [comment]
> @lpalashevski @danielaotelea can someone test that this at least now logs an issue and raises a RepositoryErrorException in such scenarios as documented in #3903 ?

@cmgrote we will pick it up with the regular regression testing we do but most probably will not be reproduced due to the test data we use. Hope that somewhere next week we will check with the product team to test it on the environment where this was originally discovered.

**Comment by lpalashevski** [comment]
We cannot merge this change do to subject-area FVT failing build checks with following exeption:


`Error:    RelationshipsIT.testRelationships:17 Unexpected exception thrown: org.odpi.openmetadata.frameworks.connectors.ffdc.PropertyServerException: OMAG-COMMON-400-016 An unexpected org.odpi.openmetadata.accessservices.subjectarea.handlers.SubjectAreaRelationshipHandler exception was caught by restore for Synonym; error message was OMAG-REPOSITORY-HANDLER-500-001 An unexpected error org.odpi.openmetadata.repositoryservices.ffdc.exception.RelationshipNotKnownException was returned to restoreRelationship(deprecated) by the metadata server during restoreSynonym request for open metadata access service Subject Area OMAS on server serverinmem; message was OMRS-REPOSITORY-404-012 The Synonym relationship identified with guid 14e101ed-bf54-4c99-912e-c94b862a3f4f passed on the getRelationship call is soft-deleted in the open metadata repository serverinmem`

This is the case for both graph and in-mem repo tests. 
Not completely sure but looks like the test should be aligned with the new validation logic for relationships. 
@davidradl can you please check?

**Comment by davidradl** [comment]
> Subject-area FVT failing build checks with following exception:
> 
> `Error: RelationshipsIT.testRelationships:17 Unexpected exception thrown: org.odpi.openmetadata.frameworks.connectors.ffdc.PropertyServerException: OMAG-COMMON-400-016 An unexpected org.odpi.openmetadata.accessservices.subjectarea.handlers.SubjectAreaRelationshipHandler exception was caught by restore for Synonym; error message was OMAG-REPOSITORY-HANDLER-500-001 An unexpected error org.odpi.openmetadata.repositoryservices.ffdc.exception.RelationshipNotKnownException was returned to restoreRelationship(deprecated) by the metadata server during restoreSynonym request for open metadata access service Subject Area OMAS on server serverinmem; message was OMRS-REPOSITORY-404-012 The Synonym relationship identified with guid 14e101ed-bf54-4c99-912e-c94b862a3f4f passed on the getRelationship call is soft-deleted in the open metadata repository serverinmem`
> 
> This is the case for both graph and in-mem repo tests.
> 
> Not completely sure but looks like the test should be aligned with the new validation logic for relationships.
> @davidradl can you please check?

Sure @lpalashevski I will check it out.

**Comment by davidradl** [comment]
From the error message, it is complaining that a restore is being attempted on a soft deleted relationship. But the only relationships that can be restored are those that are soft deleted. I can see that line 5374 in LocalOMRSMetadataCollection  is called while processing restoreRelationship. 
`    Relationship relationship = realMetadataCollection.getRelationship(userId, deletedRelationshipGUID);`
The getRelationship validation checks that the guid is associated with a non-deleted relationship. 

In the case of restore it needs to check that the relationship is deleted.

**Comment by davidradl** [comment]
@cmgrote drawing your attention to my comment above. It seems that this getRelationship validation breaks restoreRelationship, which is incorrectly using getRelationship (where your new validation is). An easy fix would be to remove the getRelationships call from restoreRelationship, a better fix would be for LocalOMRSMetadataCollection restoreRelationship to validate that the passed guid is known and soft deleted (restoreRelationship processing may already do this later) prior to calling the restoreRelationship implementation.

**Comment by cmgrote** [comment]
I'll go for the simple fix (for now) as this most closely matches the current behaviour...  Since `getRelationship` only operates on active relationships, I think to do a validation against any deleted relationship would require running a search (eg. `findRelationshipsByProperty`) which would be a (potentially much) more expensive operation than what we're doing elsewhere...

**Comment by cmgrote** [comment]
@lpalashevski can you re-test in the scenario that was failing previously?  At least the PR should build and pass all FVTs again now.

**Comment by lpalashevski** [comment]
> @lpalashevski can you re-test in the scenario that was failing previously? At least the PR should build and pass all FVTs again now.

Yes we will. Can we merge it to master already once the checks finish?

File: open-metadata-implementation/repository-services/repository-services-implementation/src/main/java/org/odpi/openmetadata/repositoryservices/localrepository/repositoryconnector/LocalOMRSMetadataCollection.java [hunk 1]
```diff
@@ -5368,19 +5368,11 @@ public Relationship restoreRelationship(String    userId,
...
-         */
-        Relationship relationship = realMetadataCollection.getRelationship(userId, deletedRelationshipGUID);
-        repositoryValidator.validateRelationshipFromStore(repositoryName, deletedRelationshipGUID, relationship, methodName);
-
         /*
          * Restore relationship
          */
         Relationship   newRelationship = realMetadataCollection.restoreRelationship(userId, deletedRelationshipGUID);
 
-        repositoryValidator.validateRelationshipFromStore(repositoryName, deletedRelationshipGUID, newRelationship, methodName);
```
Comment (davidradl) [inline open-metadata-implementation/repository-services/repository-services-implementation/src/main/java/org/odpi/openmetadata/repositoryservices/localrepository/repositoryconnector/LocalOMRSMetadataCollection.java]: This one should be ok - we are validating the relationship is not deleted after it has been restored.

Comment (lpalashevski) [reply open-metadata-implementation/repository-services/repository-services-implementation/src/main/java/org/odpi/openmetadata/repositoryservices/localrepository/repositoryconnector/LocalOMRSMetadataCollection.java]: Apologies for merging this quickly, I am trying to get it in master and include it as part of the lineage testing. Can we fix it forward with new PR putting it back?

**Pull request merged**